File: JavaScript/Trees/sortedArrayToBST.js
/**
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 * 
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    return helper(nums, 0, nums.length-1);
};

/**
 * @param {number[]} nums
 * @param {number} start
 * @param {number} end
 * @return {TreeNode}
 */
var helper = function(nums, start, end){
    if(start>end) return null;
    var mid = Math.floor((start + end) / 2);
    
    var node = new TreeNode(nums[mid]);
    
    node.left = helper(nums, start, mid - 1);
    node.right = helper(nums, mid + 1, end);
    
    return node;
};

/*
Notes: 

1. initialize start = 0 and end = array.length - 1
2. find mididx
3. Create a tree node with mid as root
4. Set node.left with recursive call on left subarray
5. Set node.right with recursive call on right subarray

Tested at https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
Completed with O(n) time complexity and O(1) space complexity

*/

File: Java/Trees/sortedArrayToBST.java
/**
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 *
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return helper(nums, 0, nums.length-1);
    }
    
    private TreeNode helper(int[] nums, int startIdx, int endIdx){
        if(startIdx > endIdx) return null;
        
        int midIdx = (startIdx + endIdx)/2;
        TreeNode node = new TreeNode(nums[midIdx]);
        
        node.left = helper(nums, startIdx, midIdx - 1);
        node.right = helper(nums, midIdx + 1, endIdx);
        
        return node;
    }
}
/*
Notes: 

1. initialize start = 0 and end = array.length - 1
2. find mididx
3. Create a tree node with mid as root
4. Set node.left with recursive call on left subarray
5. Set node.right with recursive call on right subarray

Tested at https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
Completed with O(n) time complexity and O(1) space complexity

*/